Cython/Compiler/Dataclass.py

```diff
@@ -0,0 +1,611 @@
+            return self._make_string()
+
+    def _make_string(self):
+        from .AutoDocTransforms import AnnotationWriter
+        writer = AnnotationWriter(description="Dataclass field")
+        string = writer.write(self.arg)
+        return ExprNodes.StringNode(self.pos, value=EncodedString(string))
+
+    def generate_evaluation_code(self, code):
+        return self.arg.generate_evaluation_code(code)
+
+
+def _setup_dataclass_fields(node, fields, dataclass_module):
+    from .ExprNodes import (AttributeNode, TupleNode, NameNode,
+                            GeneralCallNode, DictNode,
+                            IdentifierStringNode, BoolNode, DictItemNode,
+                            CloneNode)
+
+    # For defaults and default_factories containing things like lambda,
+    # they're already declared in the class scope, and it creates a big
+    # problem if multiple copies are floating around in both the __init__
+    # function, and in the __dataclass_fields__ structure.
+    # Therefore, create module-level constants holding these values and
+    # pass those around instead
+    variables_assignment_stats = []
+    for name, field in fields.items():
+        for attrname in [ "default", "default_factory" ]:
+            f_def = getattr(field, attrname)
+            if f_def is MISSING or f_def.is_literal or f_def.is_name:
+                # some simple cases where we don't need to set up
+                # the variable as a module-level constant
+                continue
+            global_scope = node.scope.global_scope()
+            module_field_name = global_scope.mangle(global_scope.mangle(
+                                    Naming.dataclass_field_default_cname,
+                                    node.class_name), name)
+            # create an entry in the global scope for this variable to live
+            nn = NameNode(f_def.pos, name=EncodedString(module_field_name))
+            nn.entry = global_scope.declare_var(nn.name, type=f_def.type or PyrexTypes.unspecified_type,
+                                                pos=f_def.pos, cname=nn.name, is_cdef=1)
+            # replace the field so that future users just receive the namenode
+            setattr(field, attrname, nn)
+
+            variables_assignment_stats.append(
+                Nodes.SingleAssignmentNode(f_def.pos,
+                                           lhs = nn,
+                                           rhs = DontAnalyseDeclarationsNode(f_def.pos, arg=f_def)))
+
+    placeholders = {}
+    field_func = AttributeNode(node.pos, obj = dataclass_module,
+                                    attribute=EncodedString("field"))
+    dc_fields = DictNode(node.pos, key_value_pairs=[])
+    dc_fields_namevalue_assignments = []
+    for name, field in fields.items():
+        placeholder_name = "PLACEHOLDER_%s" % name
+        placeholders[placeholder_name] = GetTypeNode(node.scope.entries[name])
```

Comment:
Can't this be done at a later stage, where we already _have_ the type available? And, is this really all that different from `typeof()`?

Reply:
The difference/complication between this and `typeof` is that `typeof` always returns a string, while I'm trying to return a Python type object if at all possible (to match what goes in `__dataclass_fields__` when generated by the standard library). It isn't always possible but I was trying to match what's generated by the standard library as closely as I could.

I think the idea for doing it here was to try to have (almost) a single transformation which adds all the methods and attributes that make a class a dataclass. After that it can just pass through the normal Cython pipeline as normal.